elfutils: enable debuginfod support

symphorien opened the pull request:
elfutils contains several libraries, and when debuginfod support is enabled, only libdebuginfod.so depends on libarchive and curl, among others, leading to a large closure size increase.

To mitigate this effect, libdebuginfod.so is moved to a separate output. libarchive and libcurl remain in the build closure, but derivations which depend on, say, libelf, but not libdebuginfod only retain a reference to the out output and not the libdebuginfod output and their runtime closure size does not increase.

Enabling libdebuginfod by default allows to use nixseparatedebuginfod without recompiling gdb and perf, mostly.
This admittedly complex contraption ensures we don't get two copies of libelf (one for systemd withouth debuginfod support, and one for debuginfod enabled gdb) in this case.

The bin output contains build tools, and since curl and libarchive are always in the build closure we dont move debuginfod and debuginfod-find to a separate output.

tested by rebuilding
/nix/store/axqgmkgmhy2rzlbffglhjv1npy5ni0ra-gdb-13.1
/nix/store/73cbnzc96xir60xff5qblhvw6fb7n309-perf-linux-6.1.21
/nix/store/nk7lla63w1pbzyavgz7hcvdhmrd9qgqy-systemtap-6.1.21-4.8
/nix/store/bhss1whsh67b0wa1mj5h3z1p4nwndxd4-valgrind-3.20.0
/nix/store/v2hlzj0gbnycj7ssznvjvhw5bxwxsbmm-hotspot-1.4.1
/nix/store/7cnyh97201zk780zwbh1sh913wflsgjp-heaptrack-1.4.0
/nix/store/1893dw0r67hh6iqbjdhwbpb8dsc1i8iy-networkmanager-1.40.12
/nix/store/xc5yfb7v042ypmbwwswpsq91jwapls4k-dwz-0.14
/nix/store/vfyvlmrnns4k0dppd6fnddkxc1s3dkck-linux-6.1.21
and checking which of these keep the debuginfod reference.

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

symphorien commented:
@ofborg eval

trofi commented on pkgs/development/tools/misc/elfutils/default.nix:
```diff
@@ -48,7 +48,7 @@ stdenv.mkDerivation rec {
   '';
 
-  outputs = [ "bin" "dev" "out" "man" ];
+  outputs = [ "bin" "dev" "out" "man" ] ++ lib.optional enableDebuginfod "libdebuginfod";
```
I would suggest it to move to existing `$bin`, not the new `$libdebuginfo`. `bin/debuginfod` depends on `libdebuginfo.so` anyways. There seems to be no reason to keep a new snowflake derivation output if nothing uses it.

symphorien replied:
most debug tools (gdb, perf, ...) depend on libdebuginfod.so but not on debuginfod itself (the lib is a client while debuginfod is a server implem), so moving debuginfod to $bin seems wasteful from a closure size point of view.

trofi commented on pkgs/development/tools/misc/elfutils/default.nix:
```diff
@@ -77,15 +77,43 @@ stdenv.mkDerivation rec {
+      mv -v $i $libdebuginfod/lib/
+    done
+
+    # fix the executables depending on it
+    for i in $bin/bin/debuginfod*; do
+      origRpath="$(patchelf --print-rpath "$i")"
+      patchelf --set-rpath "$libdebuginfod/lib:$origRpath" "$i"
+    done
```
The need for `patchelf` to keep the package running always irks me: it always has a potential missing some reference to patch. Why not keep it in `$lib` instead? `elfutils` is a toolchain-style package I would say. It's OK to have a few extra dependencies. Looking at the individual deps you mention:

- `libarchive` seems to pull in extra 2MB: 800K of libarchive itself and 1.2MB of libxml2 (the rest looks like baseline packages)
- `libcurl` seems to pull in extra 5.2MB: curl itself (800KB), libktb5 (2.2MB), brotli (1.7MB) (the rest looks like baseline packages)

I would say these are benign and worth pulling in unconditionally. The `curl` one is especially weak I would say as `curl` is already present in a typical system.

WDYT of just installing `debuginfod` by default?

symphorien replied:
> WDYT of just installing debuginfod by default?

tbh I expected that people would oppose that, given that elfutils is in the closure of approximately everything through systemd (udev), hence the extra work I did here. If you don't mind I'll wait for a second opinion before going in this direction.

> The need for patchelf to keep the package running always irks me: it always has a potential missing some reference to patch.

there are install tests, and they catch the regression if one remove the patchelf invocation, so that seems unlikely.



To give actual numbers: elfutils.bin grows from 47MB to 73MB in closure size by enabling debuginfod support, and the exact list of added dependencies is sqlite, curl, libmicrohttpd, libarchive. It is true that systemd depends on curl and libmicrohttpd, and that nix depends the remaining two, so we will likely have them in most nixos closures.

trofi reviewed (changes requested):
LGTM overall. Added one suggestion not to manually split out the dependency.

symphorien committed:
elfutils: enable debuginfod support

most reverse dependencies are debuggers which benefit from this support,
plus systemd which already depends on half the additional
depenendencies.

symphorien commented:
I changed the pull request to just setting enableDebuginfod to true without patchelf hacks.

IvarWithoutBones commented:
This PR breaks the `gdb` build on `x86_64-darwin` (and `aarch64-darwin` assuming that worked before), as `elfutils` only supports `platforms.linux`. I have confirmed that manually disabling `gdb.enableDebuginfod` with an override fixes this problem, unfortunately people on darwin now have to manually do as i did and compile gdb from source.

It would be nice if we could only enable `enableDebuginfod` by default on platforms that actually support it.

trofi commented:
Yeah, conditionally enabling in `gdb` sounds very reasonable. Let me try to add a trivial solution:  https://github.com/NixOS/nixpkgs/pull/228697

Can you give it a try?
